### PR TITLE
Ifpack2: add option to use stream interfaces of KK SPILUK and KK SPTRSV in RILUK and local triangular solver

### PR DESCRIPTION
**Pull request by vqd8a**
@trilinos/ifpack2 

## Motivation

This PR is to add a code path of using stream interfaces of KK SPILUK and KK SPTRSV in Ifpack2 RILUK and local triangular solver, respectively. Basically, a domain on each MPI process is divided to some sub-domains, each of which is handled by SPILUK and SPTRSV on a stream. These streams run concurrently.

## Testing

Tested with ```Ifpack2_tif_belos``` tester and Sierra T/F matrix ```abnormal_system_surrogate_with_gpu_rowmap-energy_fast_ilu_solver-np-16-time-292.78-solve_count-0``` on 16 MPI processes (4 MPI processes per node, V100 GPUs)
| Streams per MPI process | No stream | 2 | 4 | 9 |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Init preconditioner (seconds) | 6.35349 | 6.08481 | 5.39639 | 3.66621 |
| Compute preconditioner (seconds) | 14.9215 | 11.2069 | 8.51655 | 4.81502 |
| Belos solve (seconds) | 5.56272 | 2.35 | 2.23373 | 1.91015 |
| Total time (seconds) | 26.8475 | 19.6511 | 16.1841 | 10.4001 |
| Number of iterations | 30 | 37 | 53 | 66 |

### CODE REVIEW

**Comment by vqd8a** [comment]
@jhux2 @vbrunini @srajama1 
Could you please review this?

File: packages/ifpack2/src/Ifpack2_RILUK_def.hpp [hunk 1]
```diff
@@ -374,6 +406,12 @@ setParameters (const Teuchos::ParameterList& params)
   else {
     this->isKokkosKernelsSpiluk_ = false;
   }
+
+  {
+    const std::string paramName ("fact: kspiluk number-of-streams");
+    getParamTryingTypes<int, int, global_ordinal_type, double, float>
```
Comment (jhux2) [inline packages/ifpack2/src/Ifpack2_RILUK_def.hpp:412]:
```suggestion
    getParamTryingTypes<int, int, global_ordinal_type>
```

**Review by jhux2 — CHANGES_REQUESTED** [review]
@vqd8a I had a minor comment, otherwise this looks good.  Could you also add the option and a short description to the Ifpack2 user manual in `packages/ifpack2/doc/UsersGuide`?

**Review by srajama1 — APPROVED** [review]
@vqd8a I am approving this. But from my limited knowledge of Ifpack2 design these matrices for each stream are being created as Tpetra matrices. Is this correct? I believe these are two heavy weight for local matrices that are callable in a stream. I believe a redisgn of Ifpack2 is needed.

File: packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp [hunk 2]
```diff
@@ -310,6 +310,8 @@ void LocalSparseTriangularSolver<MatrixType>::initializeState ()
   computeTime_ = 0.0;
   applyTime_ = 0.0;
   isKokkosKernelsSptrsv_ = false;
+  isKokkosKernelsStream_ = false;
+  num_streams_ = 0;
```
[the remark this answers is not in the document]
Comment (vqd8a) [reply packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp:314]: @srajama1 Yes, I want use ```num_streams_ = 0``` for the original code path.

File: packages/ifpack2/src/Ifpack2_RILUK_def.hpp (hunk 1 above)
Comment (vqd8a) [reply packages/ifpack2/src/Ifpack2_RILUK_def.hpp]: I committed your suggestion.

**Comment by vqd8a** [comment]
> @vqd8a I am approving this. But from my limited knowledge of Ifpack2 design these matrices for each stream are being created as Tpetra matrices. Is this correct? I believe these are two heavy weight for local matrices that are callable in a stream. I believe a redisgn of Ifpack2 is needed.

@srajama1 Thanks for approving it. Yes, for now these matrices on streams are Tpetra matrices. I agree we should redesign this. I will address it in another PR.

**Comment by vqd8a** [comment]
> @vqd8a I had a minor comment, otherwise this looks good. Could you also add the option and a short description to the Ifpack2 user manual in `packages/ifpack2/doc/UsersGuide`?

@jhux2 Thanks for reviewing it. I just want to make sure this. Is `packages/ifpack2/doc/UsersGuide/options.tex` the file I should modify?

**Comment by jhux2** [comment]
> I just want to make sure this. Is `packages/ifpack2/doc/UsersGuide/options.tex` the file I should modify?

That's right.  If there are other SPILUK options you've added that aren't documented here, would you mind adding them?  Thank you.

**Comment by vqd8a** [comment]
> > I just want to make sure this. Is `packages/ifpack2/doc/UsersGuide/options.tex` the file I should modify?
> 
> That's right. If there are other SPILUK options you've added that aren't documented here, would you mind adding them? Thank you.

@jhux2 I have added options for the factorization type, trisolver type and whether using streams to the Ifpack2 userguide.

**Review by jhux2 — DISMISSED** [review]
lgtm, thank you

**Review by jhux2 — APPROVED** [review]
This looks fine to me.

**Pull request merged**